// posthog-js/lib/src/extensions/sentry-integration.js
/**
 * Integrate Sentry with PostHog. This will add a direct link to the person in Sentry, and an $exception event in PostHog
 *
 * ### Usage
 *
 *     Sentry.init({
 *          dsn: 'https://example',
 *          integrations: [
 *              new posthog.SentryIntegration(posthog)
 *          ]
 *     })
 *
 * @param {Object} [posthog] The posthog object
 * @param {string} [organization] Optional: The Sentry organization, used to send a direct link from PostHog to Sentry
 * @param {Number} [projectId] Optional: The Sentry project id, used to send a direct link from PostHog to Sentry
 * @param {string} [prefix] Optional: Url of a self-hosted sentry instance (default: https://sentry.io/organizations/)
 */
// levels copied from Sentry to avoid relying on a frequently changing @sentry/types dependency
// but provided as an array of literal types, so we can constrain the level below
var severityLevels = ['fatal', 'error', 'warning', 'log', 'info', 'debug'];
var SentryIntegration = /** @class */ (function () {
    function SentryIntegration(_posthog, organization, projectId, prefix, 
    /**
     * By default, only errors are sent to PostHog. You can set this to '*' to send all events.
     * Or to an error of SeverityLevel to only send events matching the provided levels.
     * e.g. ['error', 'fatal'] to send only errors and fatals
     * e.g. ['error'] to send only errors -- the default when omitted
     * e.g. '*' to send all events
     */
    severityAllowList) {
        if (severityAllowList === void 0) { severityAllowList = ['error']; }
        // setupOnce gets called by Sentry when it intializes the plugin
        this.name = 'posthog-js';
        this.setupOnce = function (addGlobalEventProcessor) {
            addGlobalEventProcessor(function (event) {
                var _a, _b, _c, _d, _e;
                var shouldProcessLevel = severityAllowList === '*' || severityAllowList.includes(event.level);
                if (!shouldProcessLevel || !_posthog.__loaded)
                    return event;
                if (!event.tags)
                    event.tags = {};
                var personUrl = _posthog.requestRouter.endpointFor('ui', "/project/".concat(_posthog.config.token, "/person/").concat(_posthog.get_distinct_id()));
                event.tags['PostHog Person URL'] = personUrl;
                if (_posthog.sessionRecordingStarted()) {
                    event.tags['PostHog Recording URL'] = _posthog.get_session_replay_url({ withTimestamp: true });
                }
                var exceptions = ((_a = event.exception) === null || _a === void 0 ? void 0 : _a.values) || [];
                var data = {
                    // PostHog Exception Properties,
                    $exception_message: ((_b = exceptions[0]) === null || _b === void 0 ? void 0 : _b.value) || event.message,
                    $exception_type: (_c = exceptions[0]) === null || _c === void 0 ? void 0 : _c.type,
                    $exception_personURL: personUrl,
                    // Sentry Exception Properties
                    $sentry_event_id: event.event_id,
                    $sentry_exception: event.exception,
                    $sentry_exception_message: ((_d = exceptions[0]) === null || _d === void 0 ? void 0 : _d.value) || event.message,
                    $sentry_exception_type: (_e = exceptions[0]) === null || _e === void 0 ? void 0 : _e.type,
                    $sentry_tags: event.tags,
                    $level: event.level,
                };
                if (organization && projectId)
                    data['$sentry_url'] =
                        (prefix || 'https://sentry.io/organizations/') +
                            organization +
                            '/issues/?project=' +
                            projectId +
                            '&query=' +
                            event.event_id;
                _posthog.capture('$exception', data);
                return event;
            });
        };
    }
    return SentryIntegration;
}());
export { SentryIntegration };
